apis/dscinitialization/v1/serverless_types.go

```diff
@@ -0,0 +1,68 @@
+package v1
+
+import operatorv1 "github.com/openshift/api/operator/v1"
+
+// ServerlessSpec configures KNative components used in Open Data Hub. Specifically,
+// KNative is used to enable single model serving (KServe).
+type ServerlessSpec struct {
+	// +kubebuilder:validation:Enum=Managed;Removed
+	// +kubebuilder:default=Removed
+	ManagementState operatorv1.ManagementState `json:"managementState,omitempty"`
+	// Serving configures the KNative-Serving stack used for model serving. A Service
+	// Mesh (Istio) is prerequisite, since it is used as networking layer.
+	Serving ServingSpec `json:"serving,omitempty"`
+}
+
+// ServingSpec specifies the configuration for the KNative Serving components and their
+// bindings with the Service Mesh.
+type ServingSpec struct {
+	// Name specifies the name of the KNativeServing resource that is going to be
+	// created to instruct the KNative Operator to deploy KNative serving components.
+	// +kubebuilder:default=knative-serving
+	Name string `json:"name,omitempty"`
+	// Namespace specifies the namespace where the KNativeServing resource is going
+	// to be created.
+	// +kubebuilder:default=knative-serving
+	Namespace string `json:"namespace,omitempty"`
+	// LocalGatewayServiceName allows to customize the name of the Kubernetes Service that
+	// is going to be created for intra-cluster requests. The service is created in the
+	// Service Mesh namespace.
+	// +kubebuilder:default=knative-local-gateway
+	LocalGatewayServiceName string `json:"localGatewayServiceName,omitempty"`
+	// IngressGateway allows to customize some parameters for the Istio Ingress Gateway
+	// that is bound to KNative-Serving.
+	IngressGateway IngressGatewaySpec `json:"ingressGateway,omitempty"`
+}
+
+// IngressGatewaySpec represents the configuration of the KNative Ingress Gateway.
+type IngressGatewaySpec struct {
+	// GatewaySelector specifies the label selector to choose the Istio Ingress Gateway to use
+	// for intercepting incoming requests. If unset, the selector knative=ingressgateway is used.
+	// GatewaySelector map[string]string `json:"selector,omitempty"`
+
+	// Domain specifies the DNS name for intercepting ingress requests coming from
+	// outside the cluster. Most likely, you will want to use a wildcard name,
+	// like *.example.com. If not set, the domain of the OpenShift Ingress is used.
+	// If you choose to generate a certificate, this is the domain used for the certificate request.
+	Domain string `json:"domain,omitempty"`
+	// Certificate specifies configuration about the location of the TLS certificate and
+	// if a certificate would be generated.
+	Certificate CertificateSpec `json:"certificate,omitempty"`
```

Comment:
I've been trying to figure out if we can have a default defined here instead of a `.SecretName` field of the struct itself, but clearly controller-gen is not doing it right. I might spend some time trying to provide a fix for it, but till then we need to figure out how to address a need to use `Certificate` struct in more than one place. This is also needed for Service Mesh setup.

Reply:
It is probably not that easy, since this default is mirrored into the CRD spec.

Reply:
I was hoping setting ``//+kubebuilder:default={SecretName: "default-name"` would work, as in theory it should generate proper `default` property, but it clearly doesn't at the moment. I am not sure how to proceed with this, as I will need the same field for the ServiceMesh config I believe. Perhaps not setting a default on a struct using `kubebuilder` directives, but handling it in the code itself? WDYT?

I am keen to get to the bottom of this problem, but time-wise I don't know when this can be done.

Reply:
In general, I'm happy with blank defaults, and state what would be the default in docs (here, the CRD description), since that reduces verbosity of the default YAML resource.

Reply:
Ok, let's do it this way!

Reply:
Changed to be blank, and set the default programmatically.